lib/ecto_cellar.ex

```diff
@@ -38,28 +39,90 @@ defmodule EctoCellar do
   @doc """
   Like store/2, except that if the record is invalid, raises an exception.
   """
-  @spec store!(struct(), options) :: struct()
+  @spec store!(Ecto.Schema.t() | Ecto.Changeset.t(), options) :: Ecto.Schema.t()
   def store!(%mod{} = model, opts \\ []) do
-    model_id = if id = Map.fetch!(model, id_type(opts)), do: to_string(id)
-
-    _version =
-      Version.create!(
-        %{
-          model_name: mod |> inspect(),
-          model_id: model_id,
-          model_inserted_at: model.inserted_at,
-          version: model |> cast_format_map |> Jason.encode!()
-        },
-        repo(opts)
-      )
+    Version.create!(
+      %{
+        model_name: mod |> inspect(),
+        model_id: model_id(model, opts),
+        model_inserted_at: model.inserted_at,
+        version: model |> cast_format_map |> Jason.encode!()
+      },
+      repo(opts)
+    )
 
     model
   end
 
+  @doc """
+  Inserts given model（or changeset） and stores the changes at that time in the cellar.
+   - options: EctoCellar.options()
+   - insert_or_update_opts: options for Ecto.Repo.insert/2
+  """
+  @spec insert_and_store(Ecto.Schema.t() | Ecto.Changeset.t(), options, Keyword.t()) ::
+          {:ok, Ecto.Schema.t()} | {:error, Ecto.Changeset.t()}
+  def insert_and_store(changeset, opts \\ [], insert_opts \\ []) do
+    Multi.new()
+    |> Multi.run(:model, fn _repo, _ -> repo(opts).insert(changeset, insert_opts) end)
+    |> Multi.run(:store, fn _repo, %{model: model} -> store(model, opts) end)
+    |> repo(opts).transaction()
+    |> case do
+      {:ok, %{model: model}} ->
+        {:ok, model}
+
+      error ->
+        error
+    end
+  end
```

Comment:
@mokichi
I tried wrapping some functions of Ecto.Repo.
 - insert_and_store
 - update_and_store
 - insert_or_update_and_store

I think tracking will be easier if there is something around here, but what about?

Reply:
@tashirosota Great! But I think `delete_and_store` is also needed. In addition, `upsert_store` is more concise than `insert_or_update_and_store`.

Reply:
@mokichi 
Thanks for response！
`upsert_store` is too nice！ I'll rename it soon！

> delete_and_store is also needed.

The stored mechanism of Ecto is like this.

 1. creates and stores to versions table.
 2. updates and stores to versions table.

In each step, the state is stored after issuing the insert of the model.
In other words, if the model itself does not exist with delete, it is difficult to store them.

Do you have any good ideas?

Reply:
Maybe I have to prepare a table to record the deletions.

Reply:
nnn

Reply:
Maybe it's enough to have this?
```elixir
  def delete_store(changeset, opts \\ [], delete_opts \\ []),
    do: do_wrap_func(changeset, opts, delete_opts, :delete)

  defp do_wrap_func(changeset, celler_opts, ecto_opts, func_atom) do
    Multi.new()
    |> Multi.run(:schema, fn _repo, _ ->
      apply(repo(celler_opts), func_atom, [changeset, ecto_opts])
    end)
    |> Multi.run(:store, fn _repo, %{schema: schema} -> store(schema, celler_opts) end)
    |> repo(celler_opts).transaction()
    |> case do
      {:ok, %{schema: schema}} ->
        {:ok, schema}

      error ->
        error
    end
  end
```

Reply:
https://github.com/tashirosota/ecto_cellar/pull/14/files